ReSharper.FSharp/src/FSharp/FSharp.Common/src/Checker/ScriptFcsProjectProvider.fs

```diff
@@ -132,73 +135,70 @@ type ScriptFcsProjectProvider(lifetime: Lifetime, logger: ILogger, checkerServic
-                        |> Option.map (fun options ->
-                            let parsingOptions = 
-                                { FSharpParsingOptions.Default with
-                                    SourceFiles = [| path.FullPath |]
-                                    ConditionalDefines = ImplicitDefines.scriptDefines
-                                    IsInteractive = true
-                                    IsExe = true }
-
-                            let indices = Dictionary()
-
-                            { OutputPath = path
-                              ProjectOptions = options
-                              ParsingOptions = parsingOptions
-                              FileIndices = indices
-                              ImplementationFilesWithSignatures = EmptySet.Instance
-                              ReferencedModules = EmptySet.Instance }
-                        )
-
-                    match oldOptions, newOptions with
-                    | Some oldOptions, Some newOptions ->
-                        let areEqualForChecking (options1: FSharpProjectOptions) (options2: FSharpProjectOptions) =
-                            let arrayEq a1 a2 =
-                                Array.length a1 = Array.length a2 && Array.forall2 (=) a1 a2
-
-                            arrayEq options1.OtherOptions options2.OtherOptions &&
-                            arrayEq options1.SourceFiles options2.SourceFiles
-
-                        if not (areEqualForChecking oldOptions.ProjectOptions newOptions) then
-                            optionsUpdated.Fire((path, newOptions))
-
-                    | _, Some newOptions ->
+                if cts.IsCancellationRequested then () else
+                let newOptions = getOptionsImpl path source //TODO: do not block?
+                let oldOptions = oldOptionsExistence |> Option.bind id
+
+                let fcsProject =
+                    newOptions //TODO: use createFcsProject?
+                    |> Option.map (fun options ->
+                        let parsingOptions =
+                            { FSharpParsingOptions.Default with
+                                SourceFiles = [| path.FullPath |]
+                                ConditionalDefines = ImplicitDefines.scriptDefines
+                                IsInteractive = true
+                                IsExe = true }
+
+                        let indices = Dictionary()
+
+                        { OutputPath = path
+                          ProjectOptions = options
+                          ParsingOptions = parsingOptions
+                          FileIndices = indices
+                          ImplementationFilesWithSignatures = EmptySet.Instance
+                          ReferencedModules = EmptySet.Instance }
+                    )
+                if cts.IsCancellationRequested then () else //TODO: allow stale results?
+                scriptFcsProjects[path] <- fcsProject
```

Comment:
My recommendation is to use `LifetimeDefinition` instead of CTS and rely on `lifetime.Execute()` (or `ExecuteIfAlive`) in this place.